core/src/contracts/stablex_contract/search_batches.rs

```diff
@@ -0,0 +1,202 @@
+// NOTE: Required for automock.
+#![cfg_attr(test, allow(clippy::ptr_arg))]
+
+use anyhow::{anyhow, Result};
+use ethcontract::{prelude::Web3, transport::DynTransport, web3::types::Block, BlockNumber, H256};
+use futures::{
+    compat::Future01CompatExt as _,
+    future::{BoxFuture, FutureExt as _},
+};
+#[cfg(test)]
+use mockall::automock;
+
+fn get_block_batch_id<T>(block: &Block<T>) -> u32 {
+    const BATCH_DURATION: u64 = 300;
+    (block.timestamp.as_u64() / BATCH_DURATION) as _
+}
+
+async fn get_block(
+    web3: &Web3<DynTransport>,
+    block_number: BlockNumber,
+) -> Result<ethcontract::web3::types::Block<H256>> {
+    web3.eth()
+        .block(block_number.into())
+        .compat()
+        .await?
+        .ok_or_else(|| anyhow!("block {:?} is missing", block_number))
+}
+
+#[cfg_attr(test, automock)]
+pub trait BatchIdRetrieving {
+    fn batch_id_from_block<'a>(&'a self, block_number: BlockNumber) -> BoxFuture<'a, Result<u32>>;
+
+    fn current_batch_id_and_block_number<'a>(&'a self) -> BoxFuture<'a, Result<(u32, u64)>>;
```

Comment:
I also don't like much the second method, I considered making a struct just for it but then I decided it would have been overkill.
I'm inclined towards keeping it as it is however. The name of the function reflects the order of the output values, and the two types are different, so `u32` is naturally mapped to batch numbers while `u64` means block numbers, I think this should be enough to avoid confusion.

Reply:
Sorry, what I meant is include a doc comment to clarify this a bit more (a `///` comment so it shows up in the tooltip in your IDE for example). I also don’t think the method should change.